App display page

prachisr opened the pull request.

Commit:
Add utility to render SVG image

Commit:
Add layout for app details

Commit:
Add layout for launcher fragment

Commit:
Add icons to launcher activity

Commit:
Use services file when app is in debug mode

Commit:
Fix lint issues

Commit:
Simplify init for services

Commit:
Update the app icons when services file is loaded

Commit:
Add test for launcher fragment

Commit:
Add separate grid adapter and test

Commit:
Refactor name

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Constants.kt:
```diff
@@ -0,0 +1,20 @@
+package org.freedombox.freedombox
+
+/* This file is part of FreedomBox.
```
Follow the license headers in the other source files. ** should be there in the first line and have the package after the license header

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/NetworkModule/AppLoader.kt:
```diff
@@ -0,0 +1,66 @@
+/** This file is part of FreedomBox.
```
start the license from the second line

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/NetworkModule/AppLoader.kt:
```diff
@@ -0,0 +1,66 @@
+import org.freedombox.freedombox.Views.Adapter.GridAdapter
+import org.json.JSONObject
+
+class AppLoader(val context: Context, val adapter: GridAdapter) {
```
Can remove the class here. Have the functions take context as a param. And for the function `getFBXApps` set the adapter from a param.

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/SVGModules/SvgDecoder.kt:
```diff
@@ -0,0 +1,42 @@
+            val svg = SVG.getFromInputStream(source)
+            return SimpleResource<SVG>(svg)
+        } catch (e: SVGParseException) {
+            throw IOException("Cannot load SVG")
```
Don't throw an exception. Use try-catch as a expression like this:
```kotlin
fun decode (...) = try {
    SimpleResource<SVG>(SVG.getFromInputStream(source))
} catch (_: SVGParseException) {
    null
}
```
And the compiler will infer the return type as `Resource<SVG>?` Note the `_` used as a variable name when we'd want to ignore the exception variable.
And as a general rule, avoid throwing exceptions anywhere and everywhere.

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/SVGModules/SvgDecoder.kt:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    override fun getId(): String {
```
can be written like this: 
```kotlin 
fun getId() = "org.freedombox.freedombox.svg" 
```

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/SVGModules/SvgDrawableTranscoder.kt:
```diff
@@ -0,0 +1,38 @@
+import com.caverock.androidsvg.SVG
+
+class SvgDrawableTranscoder: ResourceTranscoder<SVG, PictureDrawable> {
+    override fun getId(): String {
```
can be written like this: 
```kotlin 
fun getId() = "org.freedombox.freedombox.svg" 
```

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/SVGModules/SvgSoftwareLayerSetter.kt:
```diff
@@ -0,0 +1,44 @@
+        if (Build.VERSION_CODES.HONEYCOMB <= Build.VERSION.SDK_INT) {
+            view.setLayerType(ImageView.LAYER_TYPE_SOFTWARE, null)
+        }
+        return false
```
This seems to always return false. Is this Ok? (Check as I don't know this lib and could be wrong)

JosephKiranBabu replied:
Yes, this seems to always return false.

prachisr replied:
It does return false always as these are overriden functions and are supposed to return a Boolean, not sure about the significance of false value though. 
We can get rid of the compatibility code though as we do not support versions less than honeycomb.

prachisr replied:
Actaully, I can get rid of this class altogether as this is just doing some compatibility code for honeycomb(API version 11), are min SDK support is till 15.

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Utils/ImageRenderer.kt:
```diff
@@ -0,0 +1,41 @@
+                    .error(R.drawable.ic_logo)
+                    .listener(SvgSoftwareLayerSetter())
+
+    fun getImageFromUrl(url: Uri, imageView: ImageView) {
```
Rename this function to `loadImageFromUrl`

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+import android.widget.BaseAdapter
+import com.google.gson.JsonArray
+import com.google.gson.JsonObject
+import kotlinx.android.synthetic.main.card.view.*
```
Use fully qualified imports with the IDEA settings I had mentioned. Also format this file.

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+
+class GridAdapter(val context: Context,val imageRenderer: ImageRenderer): BaseAdapter() {
+
+    private var items:JsonArray = JsonArray()
```
No need of the data type here

prachisr replied:
@lispyclouds could you elaborate on this?

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+class GridAdapter(val context: Context,val imageRenderer: ImageRenderer): BaseAdapter() {
+
+    private var items:JsonArray = JsonArray()
+    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
```
This should work with View and ViewGroup as non nullable. eg: Just View not View?

prachisr replied:
Not exactly, because convertView and parent can have null values as well. Here is a more detailed explanation https://developer.android.com/reference/android/widget/Adapter.html

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+        rowView.appName.text = appDetail["label"]
+                .asJsonObject[locale.language]
+                .asString
+        print(rowView.appName.text.toString())
```
Remove this.

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+        return rowView
+    }
+
+    override fun getItem(position: Int): JsonObject {
```
can be written as:
```kotlin
fun getItem(position: Int) = items[position].asJsonObject
```

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+        return items[position].asJsonObject
+    }
+
+    override fun getItemId(position: Int): Long {
```
can be written as:
```kotlin
fun getItemId(position: Int) = items[position].hashCode().toLong()
```

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+        return items[position].hashCode().toLong()
+    }
+
+    override fun getCount(): Int {
```
can be written as:
```kotlin
fun getCount() = items.size()
```

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt:
```diff
@@ -0,0 +1,79 @@
+        return items.size()
+    }
+
+    fun setData(jsonArry: JsonArray) {
```
can be written as:
```kotlin
fun setData(jsonArry: JsonArray) = items = jsonArry
```
Also typo in jsonArry

JosephKiranBabu replied:
I'd say, the notation with braces is more readable in this context. The `=` symbol coming twice in succession can be confusing.

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Fragments/LauncherFragment.kt:
```diff
@@ -0,0 +1,57 @@
+package org.freedombox.freedombox.Views.Fragments
+
+import android.os.Bundle
+import kotlinx.android.synthetic.main.fragment_launcher.*
```
Use fully qualified imports

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Views/Fragments/LauncherFragment.kt:
```diff
@@ -0,0 +1,57 @@
+        appGrid.adapter = adapter
+
+        //TODO: Use the URL from settings once it is setup
+        AppLoader(context, adapter).getFBXApps(SERVICES_FILE, DEFAULT_FREEDOM_BOX_URL)
```
Use function directly

lispyclouds commented on app/src/test/java/org/freedombox/freedombox/View/Adapter/GridAdapterTest.kt:
```diff
@@ -0,0 +1,67 @@
+/** This file is part of FreedomBox.
+ *
```
Start from next line

lispyclouds commented on app/src/test/java/org/freedombox/freedombox/View/Adapter/GridAdapterTest.kt:
```diff
@@ -0,0 +1,67 @@
+
+import com.google.gson.JsonArray
+import com.google.gson.JsonParser
+import kotlinx.android.synthetic.main.card.view.*
```
Use fully qualified imports

lispyclouds commented on app/src/test/java/org/freedombox/freedombox/Views/Fragment/LauncherFragmentTest.kt:
```diff
@@ -0,0 +1,38 @@
+/** This file is part of FreedomBox.
```
Start from next line

lispyclouds reviewed (changes requested):
Reviewed for App display page

lispyclouds commented:
Also unit tests are failing for LauncherFragmentTest

lispyclouds commented on app/src/main/java/org/freedombox/freedombox/Constants.kt:
```diff
@@ -0,0 +1,20 @@
+ * along with FreedomBox. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+const val DEFAULT_FREEDOM_BOX_URL = "10.42.0.1"
```
Already a `val`. Can't get anymore `const` than this 🤣

Commit:
Using fully qualified import names

Commit:
Fix license headers in files

JosephKiranBabu commented on app/src/main/java/org/freedombox/freedombox/SVGModules/SvgSoftwareLayerSetter.kt:
```diff
@@ -0,0 +1,44 @@
+        if (Build.VERSION_CODES.HONEYCOMB <= Build.VERSION.SDK_INT) {
+            view.setLayerType(ImageView.LAYER_TYPE_SOFTWARE, null)
+        }
+        return false
```
This method always returns false too.

JosephKiranBabu commented on app/src/test/java/org/freedombox/freedombox/View/Adapter/GridAdapterTest.kt:
```diff
@@ -0,0 +1,67 @@
+    val applicationContext = application.applicationContext
+    private val gridAdapter = GridAdapter(applicationContext, imageRenderer = ImageRenderer(applicationContext))
+    private var jsonArray = JsonArray()
+    private val jsonObject = JsonParser().parse("{'label': {'en': 'conversations'}, 'description': " +
```
This looks awkward. Isn't there a way of creating a JsonObject from a Java HashMap?

JosephKiranBabu commented on app/src/test/java/org/freedombox/freedombox/View/Adapter/GridAdapterTest.kt:
```diff
@@ -0,0 +1,67 @@
+        gridAdapter.setData(jsonArray)
+        assertEquals(gridAdapter.count, 1)
+    }
+
```
Isn't this over-testing? Do we really need tests for getters and setters of GridAdapter class?

JosephKiranBabu commented on app/src/test/java/org/freedombox/freedombox/Views/Fragment/LauncherFragmentTest.kt:
```diff
@@ -0,0 +1,38 @@
+    @Test
+    fun viewIsGettingPopulatedWithIcons() {
+        val launcherFragment = LauncherFragment()
+        startFragment(launcherFragment)
```
This test passes whether you run `startFragment(launcherFragment)` or not. Nothing is being tested here.

JosephKiranBabu commented on app/src/main/assets/services.json:
```diff
@@ -0,0 +1,127 @@
+{
```
I don't think this file belongs in this app. We might use this in tests for loading additional apps though.

Commit:
Removed unused class SvgSoftwareLayerSetter

Commit:
Refactored functions in SVG Modules

Commit:
Pull out function and remove appLoader class

Notify to reload adapter with new items

- Refactor declaration

Add sharedPreference to store apps response

lispyclouds commented:
Ok. Should be fine as long as we are not using them.


On Aug 8, 2017 at 12:35 PM, <prachisr <notifications@github.com>> wrote:

*@prachisr* commented on this pull request.
------------------------------

In app/src/main/java/org/freedombox/freedombox/Views/Adapter/GridAdapter.kt
<https://github.com/freedombox/android-app/pull/16#discussion_r131835210>:

> +import android.net.Uri
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.BaseAdapter
+import com.google.gson.JsonArray
+import com.google.gson.JsonObject
+import kotlinx.android.synthetic.main.card.view.*
+import org.freedombox.freedombox.R
+import org.freedombox.freedombox.Utils.ImageRenderer
+import java.util.*
+
+class GridAdapter(val context: Context,val imageRenderer:
ImageRenderer): BaseAdapter() {
+
+    private var items:JsonArray = JsonArray()
+    override fun getView(position: Int, convertView: View?, parent:
ViewGroup?): View {

Not exactly, because convertView and parent can have null values as well.
Here is a more detailed explanation
https://developer.android.com/reference/android/widget/Adapter.html

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/freedombox/android-app/pull/16#discussion_r131835210>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AFWv5IjxGG7LfV-KwoUl9QNc3oOme3oiks5sWAiugaJpZM4Ou_WJ>
.

Commit:
Format json in gridAdapter test

lispyclouds commented:
Merged with minor fixes.
